// xde/build/ui/Modals.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Constants = require('./Constants');

var _NewProjectModal = require('./NewProjectModal');

var _NewProjectModal2 = _interopRequireDefault(_NewProjectModal);

var _StyleConstants = require('./StyleConstants');

var _StyleConstants2 = _interopRequireDefault(_StyleConstants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Modals = (_temp = _class = function (_React$Component) {
  (0, _inherits3.default)(Modals, _React$Component);

  function Modals() {
    (0, _classCallCheck3.default)(this, Modals);
    return (0, _possibleConstructorReturn3.default)(this, (Modals.__proto__ || (0, _getPrototypeOf2.default)(Modals)).apply(this, arguments));
  }

  (0, _createClass3.default)(Modals, [{
    key: 'render',
    value: function render() {
      if (!this.props.currentModal) {
        return null;
      }

      var el = null;
      switch (this.props.currentModal) {
        case _Constants.ModalEnum.NEW_PROJECT:
          return _react2.default.createElement(_NewProjectModal2.default, {
            onClose: this.props.onCloseModal,
            onSelectProject: this.props.onSelectProject
          });
      }

      if (el) {
        return _react2.default.createElement(
          'div',
          { style: Styles.modalOverlay },
          _react2.default.createElement(
            'div',
            { style: Styles.modalContent },
            el
          )
        );
      } else {
        return null;
      }
    }
  }]);
  return Modals;
}(_react2.default.Component), _class.propTypes = {
  currentModal: _react.PropTypes.string,
  onSelectProject: _react.PropTypes.func.isRequired,
  onCloseModal: _react.PropTypes.func.isRequired
}, _temp);
exports.default = Modals;


var Styles = {
  modalOverlay: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.8)',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
  },

  modalContent: {
    background: _StyleConstants2.default.colorBackground,
    overflow: 'auto',
    borderRadius: '2px',
    outline: 'none'
  }
};
