// xde/build/ui/Notification.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _StyleConstants = require('./StyleConstants');

var _StyleConstants2 = _interopRequireDefault(_StyleConstants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MAX_LENGTH = 150;

var Notification = (_temp2 = _class = function (_React$Component) {
  (0, _inherits3.default)(Notification, _React$Component);

  function Notification() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Notification);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Notification.__proto__ || (0, _getPrototypeOf2.default)(Notification)).call.apply(_ref, [this].concat(args))), _this), _this._onClick = function (event) {
      event.stopPropagation();
      if (_this.props.onClick) {
        _this.props.onClick();
      }
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Notification, [{
    key: 'render',
    value: function render() {
      var color = void 0;
      switch (this.props.type) {
        case 'error':
          color = _StyleConstants2.default.colorError;
          break;
        case 'warning':
          color = _StyleConstants2.default.colorWarning;
          break;
        default:
          color = _StyleConstants2.default.colorPrimary;
          break;
      }

      var message = this.props.message;
      if (message.length > MAX_LENGTH) {
        message = message.substring(0, MAX_LENGTH) + '...';
      }

      return _react2.default.createElement(
        'div',
        {
          style: (0, _extends3.default)({}, Styles.container, {
            backgroundColor: color,
            cursor: this.props.onClick ? 'pointer' : 'default'
          }), onClick: this._onClick },
        message
      );
    }
  }]);
  return Notification;
}(_react2.default.Component), _class.propTypes = {
  type: _react.PropTypes.oneOf(['success', 'error', 'warning', 'info']),
  message: _react.PropTypes.string.isRequired,
  onClick: _react.PropTypes.func
}, _temp2);
exports.default = Notification;


var Styles = {
  container: {
    color: 'white',
    padding: _StyleConstants2.default.gutterMd,
    textAlign: 'center',

    position: 'fixed',
    top: 0,
    left: 0,
    right: 0
  }
};
