// xde/build/ui/ProjectList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2, _class2, _temp4;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ProjectIcon = require('./ProjectIcon');

var _ProjectIcon2 = _interopRequireDefault(_ProjectIcon);

var _StyleConstants = require('./StyleConstants');

var _StyleConstants2 = _interopRequireDefault(_StyleConstants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var projectPropType = _react.PropTypes.shape({
  root: _react.PropTypes.string,
  name: _react.PropTypes.string,
  readableRoot: _react.PropTypes.string,
  icon: _react.PropTypes.string
});

var Project = (_temp2 = _class = function (_React$Component) {
  (0, _inherits3.default)(Project, _React$Component);

  function Project() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Project);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Project.__proto__ || (0, _getPrototypeOf2.default)(Project)).call.apply(_ref, [this].concat(args))), _this), _this._onClick = function () {
      _this.props.onSelect(_this.props.project);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Project, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        { style: Styles.project, onClick: this._onClick },
        _react2.default.createElement(_ProjectIcon2.default, { size: 32, iconUrl: this.props.project.icon }),
        _react2.default.createElement(
          'div',
          { style: Styles.projectDetails },
          _react2.default.createElement(
            'div',
            { style: Styles.projectName },
            this.props.project.name
          ),
          _react2.default.createElement(
            'div',
            { style: Styles.projectRoot },
            this.props.project.readableRoot
          )
        )
      );
    }
  }]);
  return Project;
}(_react2.default.Component), _class.propTypes = {
  project: projectPropType.isRequired,
  onSelect: _react.PropTypes.func.isRequired
}, _temp2);
var ProjectList = (_temp4 = _class2 = function (_React$Component2) {
  (0, _inherits3.default)(ProjectList, _React$Component2);

  function ProjectList() {
    var _ref2;

    var _temp3, _this2, _ret2;

    (0, _classCallCheck3.default)(this, ProjectList);

    for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    return _ret2 = (_temp3 = (_this2 = (0, _possibleConstructorReturn3.default)(this, (_ref2 = ProjectList.__proto__ || (0, _getPrototypeOf2.default)(ProjectList)).call.apply(_ref2, [this].concat(args))), _this2), _this2._renderProject = function (project, index) {
      return _react2.default.createElement(Project, { key: index, project: project, onSelect: _this2.props.onSelect });
    }, _temp3), (0, _possibleConstructorReturn3.default)(_this2, _ret2);
  }

  (0, _createClass3.default)(ProjectList, [{
    key: 'render',
    value: function render() {
      if (this.props.projects.length === 0) {
        return _react2.default.createElement(
          'div',
          { style: Styles.container },
          _react2.default.createElement(
            'div',
            { style: Styles.placeholder },
            'No recent projects to show. Use the Project button to get started.'
          )
        );
      }
      return _react2.default.createElement(
        'div',
        { style: Styles.container },
        _react2.default.createElement(
          'div',
          { style: Styles.projectList },
          this.props.projects.map(this._renderProject)
        )
      );
    }
  }]);
  return ProjectList;
}(_react2.default.Component), _class2.propTypes = {
  projects: _react.PropTypes.arrayOf(projectPropType).isRequired,
  onSelect: _react.PropTypes.func.isRequired
}, _temp4);
exports.default = ProjectList;


var Styles = {
  container: {
    borderTop: '1px solid ' + _StyleConstants2.default.colorBorder,
    height: '100%',
    overflowY: 'auto',

    display: 'flex'
  },
  placeholder: {
    color: _StyleConstants2.default.colorSubtitle,
    fontSize: _StyleConstants2.default.fontSizeLg,
    maxWidth: 460,
    margin: 'auto'
  },
  projectList: {
    flex: 1
  },
  project: {
    cursor: 'pointer',
    margin: _StyleConstants2.default.gutterLg,

    display: 'flex',
    alignItems: 'center'
  },
  projectDetails: {
    flex: 1,
    marginLeft: _StyleConstants2.default.gutterMd
  },
  projectName: {
    color: _StyleConstants2.default.colorText,
    fontSize: _StyleConstants2.default.fontSizeLg
  },
  projectRoot: {
    color: _StyleConstants2.default.colorSubtitle,
    fontSize: _StyleConstants2.default.fontSizeMd
  }
};
