// xde/build/xde-cli.js
#!/usr/bin/env node
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

// We can't do `export default` here or else `xdeAsync` won't be defined below

var xdeAsync = function () {
  var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
    var jsonArgs;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            jsonArgs = (0, _stringify2.default)(process.argv.slice(2));
            // console.log("Running XDE with args:", jsonArgs);

            return _context.abrupt('return', _child_process2.default.spawn(executable, [electronAppRoot], {
              env: (0, _extends3.default)({}, process.env, {
                // XDE_NPM_START: 1,
                XDE_CMD_LINE_ARGS: jsonArgs,
                XDE_CMD_LINE_CWD: process.cwd()
              }),
              detached: true,
              stdio: 'ignore'
            }));

          case 2:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function xdeAsync() {
    return _ref.apply(this, arguments);
  };
}();

var _child_process = require('child_process');

var _child_process2 = _interopRequireDefault(_child_process);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _promisePrint = require('promise-print');

var _promisePrint2 = _interopRequireDefault(_promisePrint);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var executable = _path2.default.resolve(__dirname, '../node_modules/.bin/electron');
var electronAppRoot = _path2.default.resolve(__dirname, '..');exports.default = xdeAsync;


if (require.main === module) {
  (0, _promisePrint2.default)(xdeAsync().then(function () {
    process.exit(0);
  }));
}
