// xdl/build/Analytics.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setSegmentNodeKey = setSegmentNodeKey;
exports.setSegmentWebInstance = setSegmentWebInstance;
exports.setUserProperties = setUserProperties;
exports.setVersionName = setVersionName;
exports.logEvent = logEvent;

var _ip = require('ip');

var _ip2 = _interopRequireDefault(_ip);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _analyticsNode = require('analytics-node');

var _analyticsNode2 = _interopRequireDefault(_analyticsNode);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _segmentNodeInstance = void 0;

var _segmentWebInstance = void 0;
var _userId = void 0;
var _version = void 0;
var PLATFORM_TO_ANALYTICS_PLATFORM = {
  'darwin': 'Mac',
  'win32': 'Windows',
  'linux': 'Linux'
};

function setSegmentNodeKey(key) {
  _segmentNodeInstance = new _analyticsNode2.default(key);
}

function setSegmentWebInstance(instance) {
  _segmentWebInstance = instance;
}

function setUserProperties(userId, traits) {
  _userId = userId;

  if (_segmentNodeInstance) {
    _segmentNodeInstance.identify({
      userId: userId,
      traits: traits,
      context: _getContext()
    });
  }

  if (_segmentWebInstance) {
    // The Amplitude SDK isn't initialized right away, so call setVersion before every call to make sure it's actually updated.
    setVersionName(_version);

    window.analytics.identify(userId, traits, {
      context: _getContext()
    });
  }
}

function setVersionName(version) {
  _version = version;

  if (_segmentWebInstance && window.amplitude && window.amplitude.getInstance && window.amplitude.getInstance()) {
    // Segment injects amplitude into the window. Call this manually because Segment isn't passing it along.
    window.amplitude.getInstance().setVersionName(version);
  }
}

function logEvent(name) {
  var properties = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  if (_segmentNodeInstance && _userId) {
    _segmentNodeInstance.track({
      userId: _userId,
      event: name,
      properties: properties,
      context: _getContext()
    });
  }

  if (_segmentWebInstance) {
    // The Amplitude SDK isn't initialized right away, so call setVersion before every call to make sure it's actually updated.
    setVersionName(_version);

    window.analytics.track(name, properties, {
      context: _getContext()
    });
  }
}

function _getContext() {
  var platform = PLATFORM_TO_ANALYTICS_PLATFORM[_os2.default.platform()];
  var context = {
    ip: _ip2.default.address(),
    device: {
      model: platform,
      brand: platform
    },
    os: {
      name: platform,
      version: _os2.default.release()
    },
    app: {}
  };

  if (_version) {
    context.app = {
      version: _version
    };
  }

  return context;
}
