// xdl/build/Config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Env = require('./Env');

var Env = _interopRequireWildcard(_Env);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var scheme = process.env.XDL_SCHEME || 'https';

var host = process.env.XDL_HOST || 'exp.host';
var port = parseInt(process.env.XDL_PORT, 10) || null;

if (Env.isStaging()) {
  host = 'staging.exp.host';
} else if (Env.isLocal()) {
  scheme = 'http';
  host = 'localhost';
  port = 3000;
}

exports.default = {
  api: {
    scheme: scheme,
    host: host,
    port: port
  },
  ngrok: {
    authToken: '5W1bR67GNbWcXqmxZzBG1_56GezNeaX6sSRvn8npeQ8',
    authTokenPublicId: '5W1bR67GNbWcXqmxZzBG1',
    domain: 'exp.direct'
  },
  developerTool: null,
  validation: {
    reactNativeVersionWarnings: true
  },
  helpUrl: 'https://docs.getexponent.com/'
};
