// xdl/build/Logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _bunyan = require('bunyan');

var _bunyan2 = _interopRequireDefault(_bunyan);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _UserSettings = require('./UserSettings');

var _UserSettings2 = _interopRequireDefault(_UserSettings);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var logger = _bunyan2.default.createLogger({
  name: 'exponent',
  streams: [{
    level: 'debug',
    type: 'rotating-file',
    path: _path2.default.join(_UserSettings2.default.dotExponentHomeDirectory(), 'log'),
    period: '1d', // daily rotation
    count: 3 }]
});

logger.notifications = logger.child({ type: 'notifications' });
logger.global = logger.child({ type: 'global' });
logger.DEBUG = _bunyan2.default.DEBUG;
logger.INFO = _bunyan2.default.INFO;
logger.WARN = _bunyan2.default.WARN;
logger.ERROR = _bunyan2.default.ERROR;

exports.default = logger;
