// xdl/build/ProjectSettings.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPackagerOptsAsync = exports.setPackagerInfoAsync = exports.readPackagerInfoAsync = exports.setAsync = exports.readAsync = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var readAsync = exports.readAsync = function () {
  var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(projectRoot) {
    var projectSettings;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            projectSettings = void 0;
            _context.prev = 1;
            _context.next = 4;
            return projectSettingsJsonFile(projectRoot, projectSettingsFile).readAsync();

          case 4:
            projectSettings = _context.sent;
            _context.next = 12;
            break;

          case 7:
            _context.prev = 7;
            _context.t0 = _context['catch'](1);
            _context.next = 11;
            return projectSettingsJsonFile(projectRoot, projectSettingsFile).writeAsync(projectSettingsDefaults);

          case 11:
            projectSettings = _context.sent;

          case 12:

            if (projectSettings.hostType === 'ngrok') {
              // 'ngrok' is deprecated
              projectSettings.hostType = 'tunnel';
            }

            // Set defaults for any missing fields
            _lodash2.default.defaults(projectSettings, projectSettingsDefaults);
            return _context.abrupt('return', projectSettings);

          case 15:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[1, 7]]);
  }));

  return function readAsync(_x) {
    return _ref.apply(this, arguments);
  };
}();

var setAsync = exports.setAsync = function () {
  var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(projectRoot, json) {
    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.prev = 0;
            _context2.next = 3;
            return projectSettingsJsonFile(projectRoot, projectSettingsFile).mergeAsync(json, { cantReadFileDefault: projectSettingsDefaults });

          case 3:
            return _context2.abrupt('return', _context2.sent);

          case 6:
            _context2.prev = 6;
            _context2.t0 = _context2['catch'](0);
            _context2.next = 10;
            return projectSettingsJsonFile(projectRoot, packagerInfoFile).writeAsync(_lodash2.default.defaults(json, projectSettingsDefaults));

          case 10:
            return _context2.abrupt('return', _context2.sent);

          case 11:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this, [[0, 6]]);
  }));

  return function setAsync(_x2, _x3) {
    return _ref2.apply(this, arguments);
  };
}();

var readPackagerInfoAsync = exports.readPackagerInfoAsync = function () {
  var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(projectRoot) {
    return _regenerator2.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.prev = 0;
            _context3.next = 3;
            return projectSettingsJsonFile(projectRoot, packagerInfoFile).readAsync({ cantReadFileDefault: {} });

          case 3:
            return _context3.abrupt('return', _context3.sent);

          case 6:
            _context3.prev = 6;
            _context3.t0 = _context3['catch'](0);
            _context3.next = 10;
            return projectSettingsJsonFile(projectRoot, packagerInfoFile).writeAsync({});

          case 10:
            return _context3.abrupt('return', _context3.sent);

          case 11:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this, [[0, 6]]);
  }));

  return function readPackagerInfoAsync(_x4) {
    return _ref3.apply(this, arguments);
  };
}();

var setPackagerInfoAsync = exports.setPackagerInfoAsync = function () {
  var _ref4 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4(projectRoot, json) {
    return _regenerator2.default.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.prev = 0;
            _context4.next = 3;
            return projectSettingsJsonFile(projectRoot, packagerInfoFile).mergeAsync(json, { cantReadFileDefault: {} });

          case 3:
            return _context4.abrupt('return', _context4.sent);

          case 6:
            _context4.prev = 6;
            _context4.t0 = _context4['catch'](0);
            _context4.next = 10;
            return projectSettingsJsonFile(projectRoot, packagerInfoFile).writeAsync(json);

          case 10:
            return _context4.abrupt('return', _context4.sent);

          case 11:
          case 'end':
            return _context4.stop();
        }
      }
    }, _callee4, this, [[0, 6]]);
  }));

  return function setPackagerInfoAsync(_x5, _x6) {
    return _ref4.apply(this, arguments);
  };
}();

var getPackagerOptsAsync = exports.getPackagerOptsAsync = function () {
  var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee5(projectRoot) {
    var projectSettings;
    return _regenerator2.default.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.next = 2;
            return readAsync(projectRoot);

          case 2:
            projectSettings = _context5.sent;
            return _context5.abrupt('return', projectSettings);

          case 4:
          case 'end':
            return _context5.stop();
        }
      }
    }, _callee5, this);
  }));

  return function getPackagerOptsAsync(_x7) {
    return _ref5.apply(this, arguments);
  };
}();

exports.dotExponentProjectDirectory = dotExponentProjectDirectory;
exports.dotExponentProjectDirectoryExists = dotExponentProjectDirectoryExists;

require('instapromise');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _jsonFile = require('@exponent/json-file');

var _jsonFile2 = _interopRequireDefault(_jsonFile);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _mkdirp = require('mkdirp');

var _mkdirp2 = _interopRequireDefault(_mkdirp);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var projectSettingsFile = 'settings.json';
var projectSettingsDefaults = {
  hostType: 'tunnel',
  lanType: 'ip',
  dev: true,
  strict: false,
  minify: false,
  urlType: 'exp',
  urlRandomness: null
};
var packagerInfoFile = 'packager-info.json';

function projectSettingsJsonFile(projectRoot, filename) {
  return new _jsonFile2.default(_path2.default.join(dotExponentProjectDirectory(projectRoot), filename));
}

function dotExponentProjectDirectory(projectRoot) {
  var dirPath = _path2.default.join(projectRoot, '.exponent');
  try {
    // remove .exponent file if it exists, we moved to a .exponent directory
    if (_fs2.default.statSync(dirPath).isFile()) {
      _fs2.default.unlinkSync(dirPath);
    }
  } catch (e) {
    // no file or directory, continue
  }

  _mkdirp2.default.sync(dirPath);
  return dirPath;
}

function dotExponentProjectDirectoryExists(projectRoot) {
  var dirPath = _path2.default.join(projectRoot, '.exponent');
  try {
    if (_fs2.default.statSync(dirPath).isDirectory()) {
      return true;
    }
  } catch (e) {
    // file doesn't exist
  }

  return false;
}
