// xde/build/ui/toolbar/MenuSeparator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _StyleConstants = require('../StyleConstants');

var _StyleConstants2 = _interopRequireDefault(_StyleConstants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MenuSeparator = function (_React$Component) {
  (0, _inherits3.default)(MenuSeparator, _React$Component);

  function MenuSeparator() {
    (0, _classCallCheck3.default)(this, MenuSeparator);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(MenuSeparator).apply(this, arguments));
  }

  (0, _createClass3.default)(MenuSeparator, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement('div', { style: Styles.separator });
    }
  }]);
  return MenuSeparator;
}(_react2.default.Component);

exports.default = MenuSeparator;


var Styles = {
  separator: {
    borderBottom: '1px solid ' + _StyleConstants2.default.colorBorder,
    marginTop: _StyleConstants2.default.gutterMd,
    marginBottom: _StyleConstants2.default.gutterMd
  }
};
