// xde/build/ui/toolbar/MenuItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _StyleConstants = require('../StyleConstants');

var _StyleConstants2 = _interopRequireDefault(_StyleConstants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MenuItem = (_temp = _class = function (_React$Component) {
  (0, _inherits3.default)(MenuItem, _React$Component);

  function MenuItem(props, context) {
    (0, _classCallCheck3.default)(this, MenuItem);

    var _this = (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(MenuItem).call(this, props, context));

    _this._onMouseEnter = function () {
      _this.setState({ isHovered: true });
    };

    _this._onMouseLeave = function () {
      _this.setState({ isHovered: false });
    };

    _this._onClick = function (event) {
      if (_this.props.isDisabled) {
        event.stopPropagation();
      } else if (_this.props.onClick) {
        _this.props.onClick(event);
      }
    };

    _this.state = { isHovered: false };
    return _this;
  }

  (0, _createClass3.default)(MenuItem, [{
    key: '_renderFlyout',
    value: function _renderFlyout() {
      return _react2.default.createElement(
        'div',
        { style: {
            position: 'absolute',
            top: 0,

            // TODO: When we need flyouts to the left, make this customizable.
            left: 0,
            marginLeft: '100%'
          } },
        this.props.children
      );
    }
  }, {
    key: '_renderCheck',
    value: function _renderCheck() {
      // If unchecked, hold a space for the checkmark by rendering an invisible one.
      var style = this.props.checkState === 'checked' ? {} : { opacity: 0 };
      return _react2.default.createElement(
        'span',
        { style: style },
        '✓ '
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var isFlyoutMenu = !!this.props.children;
      var menuItemStyle = (0, _extends3.default)({}, Styles.menuItem, this.state.isHovered && !this.props.isDisabled ? Styles.menuItemHovered : {}, this.props.isDisabled ? Styles.menuItemDisabled : {});
      return _react2.default.createElement(
        'div',
        { style: menuItemStyle,
          onClick: this._onClick,
          onMouseEnter: this._onMouseEnter,
          onMouseLeave: this._onMouseLeave },
        this.props.checkState !== 'uncheckable' && this._renderCheck(),
        _react2.default.createElement(
          'div',
          { style: Styles.label },
          this.props.label
        ),
        this.props.shortcut && _react2.default.createElement(
          'div',
          { style: Styles.shortcut },
          this._metaKey(),
          this.props.shortcut
        ),
        isFlyoutMenu && _react2.default.createElement(
          'span',
          null,
          '►'
        ),
        isFlyoutMenu && this.state.isHovered && this._renderFlyout()
      );
    }
  }, {
    key: '_metaKey',
    value: function _metaKey() {
      if (process.platform === 'darwin') {
        return '⌘';
      } else {
        return 'Ctrl+';
      }
    }
  }]);
  return MenuItem;
}(_react2.default.Component), _class.propTypes = {
  checkState: _react.PropTypes.oneOf(['uncheckable', 'unchecked', 'checked']),

  isDisabled: _react.PropTypes.bool,
  label: _react.PropTypes.string.isRequired,
  onClick: _react.PropTypes.func,
  shortcut: _react.PropTypes.string,

  // If specified, this becomes a flyout menu.
  children: _react.PropTypes.node
}, _class.defaultProps = {
  checkState: 'uncheckable'
}, _temp);
exports.default = MenuItem;


var Styles = {
  menuItem: {
    color: _StyleConstants2.default.colorText,
    cursor: 'pointer',
    fontSize: _StyleConstants2.default.fontSizeMd,
    paddingTop: _StyleConstants2.default.gutterSm,
    paddingRight: _StyleConstants2.default.gutterMd,
    paddingBottom: _StyleConstants2.default.gutterSm,
    paddingLeft: _StyleConstants2.default.gutterMd,
    position: 'relative',
    textAlign: 'left',

    display: 'flex',
    alignItems: 'center'
  },
  menuItemHovered: {
    backgroundColor: _StyleConstants2.default.colorPrimary,
    color: 'white'
  },
  menuItemDisabled: {
    color: _StyleConstants2.default.colorSubtitle,
    cursor: 'default'
  },
  label: {
    flex: 1
  },
  shortcut: {
    color: _StyleConstants2.default.colorSubtitle,
    flex: 'none'
  }
};
